Tools/table.5
.\" ************    LibreSilicon's StdCellLibrary   *******************
.\"
.\" Organisation:   Chipforge
.\"                 Germany / European Union
.\"
.\" Profile:        Chipforge focus on fine System-on-Chip Cores in
.\"                 Verilog HDL Code which are easy understandable and
.\"                 adjustable. For further information see
.\"                         www.chipforge.org
.\"                 there are projects from small cores up to PCBs, too.
.\"
.\" File:           StdCellLib/Tools/table.5
.\"
.\" Purpose:        man 5 table - classical Manual Page
.\"
.\" ************    `groff -t -mdoc`    *******************************
.\"
.\" ///////////////////////////////////////////////////////////////////
.\"
.\" Copyright (c)   2018 - 2022 by
.\"                 chipforge - <popcorn@nospam.chipforge.org>
.\"
.\" This source file may be used and distributed without restriction
.\" provided that this copyright statement is not removed from the
.\" file and that any derivative work contains the original copyright
.\" notice and the associated disclaimer.
.\"
.\" This source is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This source is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU General Public License for more details.
.\"
.\"  (__)  You should have received a copy of the GNU General Public
.\"  oo )  License along with this program; if not, write to the
.\"  /_/|  Free Software Foundation Inc., 51 Franklin St., 5th Floor,
.\"        Boston, MA 02110-1301, USA
.\"
.\" GNU General Public License v3.0 - http://www.gnu.org/licenses/gpl-3.0.html
.\" ///////////////////////////////////////////////////////////////////
.Dd July 4, 2022
.Dt TABLE 5 "Standard Cell Library"
.Os LibreSilicon
.Sh NAME
.Nm table
.Nd description file format for Truth Tables as used by the
.Xr scarbata 1
tool.
.Sh DESCRIPTION
This is a secondary file format for the Standard Cell Library, inspired by tables from
User-defined Primitives (UDP) in Verilog (defined in IEEE Std 1394-1995 and IEEE Std 1394-2001),
and influenced by the
.Xr cell 5
file format.
.Pp
The
.Nm
file consists of newline separated ASCII records.
.Pp
Every line starts with comment, an identifier, or forms a table entry.
.Ss // 
Two slash signs mark the beginning of a comment line; subsequent characters up
to the end of the line are not interpreted by the routines which search the
file.
.Ss .cell
This dot-command is a string giving the unique cell identifier inside the
Standard Cell Library.
E.g.
.Em INV ,
.Em NAND2 ,
.Em NOR2 ,
.Em XOR2 ...
.Pp
This command is optional.
.Ss .inputs
This dot-command follows a white-space-separated list of strings (terminated by
the end of the line) giving the formal input signal names for the following table being
declared.
.Ss .outputs
This dot-command follows a white-space-separated list of strings (terminated by
the end of the line) giving the formal output signals for the cell being
delared.
.Ss .clocks
This dot-command follows a white-space-separated list of strings (terminated by
the end of the line) giving the clock signal names for the following table being declared.
This command is optional; e.g., for purely combinational circuits.
.Ss .table
This identifier following all table entry lines, until the
.Em end
identifier.
.Ss .end
This identifier indicates the end of table entry line section.
.Pp
Every table line contains rows of values in the following order:
.Bl -ohang
.It Em .outputs
A list of space-separeted Output values.
.It Em :
The colon sign devided the Output Node side of table from the following Nodes.
.It Em .clocks
Optional clock signal values, when mentioned, are followed by another
.It Em :
Colon sign.
.It Em .inputs
Behind last colon sign all Input values are listed.
.El
.Pp
Valid values for common Input, Clock or Output nodes are:
.Bl -column -offset indent ".Sy Value" ".Sy alternative" ".Sy Description"
.It Sy Value Ta Sy alternative Ta Sy Description
.It Li 0 Ta L Ta logical '0', or low
.It Li 1 Ta H Ta logical '1', or high
.It Li ? Ta -, * Ta don't care, or ignored
.It Li X Ta Ta unknown, or undefined
.El
.Pp
In addition to valid common values, on Clock nodes are also valid:
.Bl -column -offset indent ".Sy Value" ".Sy alternative" ".Sy Description"
.It Sy Value Ta Sy alternative Ta Sy Description
.It Li / Ta R, P, (01) Ta rising or positive edge
.It Li \e Ta F, N, (10) Ta falling or negative edge
.El
.Pp
In addition to valid common values, on Output nodes is also valid:
.Bl -column -offset indent ".Sy Value" ".Sy alternative" ".Sy Description"
.It Sy Value Ta Sy alternative Ta Sy Description
.It Li Z Ta Ta tri-stated, also known as hi-Z, floating, or high impedance
.El
.Sh AUTHORS
.An chipforge Aq scarbata@nospam.chipforge.org
.Sh BUGS
Not yet known.
Please do not hesitate to report your findings.
.Sh COPYRIGHT
Copyright (c) 2018-2022 Author and Contributors.
.Pp
GNU General Public License v3.0
.UR
http://www.gnu.org/licenses/gpl-3.0.html
.UE
.Sh SEE ALSO
.Xr scarbata 1
on
.UR
https://github.com/chipforge/StdCellLib
.UE
for current version and others auxilary utilities around.


Tools/exporter/table.scm
;;  ************    LibreSilicon's StdCellLibrary   *******************
;;
;;  Organisation:   Chipforge
;;                  Germany / European Union
;;
;;  Profile:        Chipforge focus on fine System-on-Chip Cores in
;;                  Verilog HDL Code which are easy understandable and
;;                  adjustable. For further information see
;;                          www.chipforge.org
;;                  there are projects from small cores up to PCBs, too.
;;
;;  File:           StdCellLib/Tools/exporter/table.scm
;;
;;  Purpose:        Scheme Module - export table dataset as library
;;
;;  ************    Revised^7 Report on Scheme (R7RS)   ***************
;;
;;  ///////////////////////////////////////////////////////////////////
;;
;;  Copyright (c)   2021 by
;;                  chipforge <popcorn@nospam.chipforge.org>
;;
;;  This source file may be used and distributed without restriction
;;  provided that this copyright statement is not removed from the
;;  file and that any derivative work contains the original copyright
;;  notice and the associated disclaimer.
;;
;;  This source is free software; you can redistribute it and/or modify
;;  it under the terms of the GNU General Public License as published by
;;  the Free Software Foundation; either version 3 of the License, or
;;  (at your option) any later version.
;;
;;  This source is distributed in the hope that it will be useful,
;;  but WITHOUT ANY WARRANTY; without even the implied warranty of
;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
;;  GNU General Public License for more details.
;;
;;   (__)  You should have received a copy of the GNU General Public
;;   oo )  License along with this program; if not, write to the
;;   /_/|  Free Software Foundation Inc., 51 Franklin St., 5th Floor,
;;         Boston, MA 02110-1301, USA
;;
;;  GNU General Public License v3.0 - http://www.gnu.org/licenses/gpl-3.0.html
;;  ///////////////////////////////////////////////////////////////////

(define-library (exporter table)
  (import (scheme base)
          (srfi 78)         ; test suite
) (export ; exporter
          exporter:dataset-table
) (begin

;;  ------------    srfi-78 test suite  -------------------------------

    ; change this switch during development only
    ; mode must be a symbol in '(off summary report-failed report)
    (check-set-mode! 'off)

;;  ------------    exporter table file -------------------------------

    (define (exporter:dataset-table table)
        "Generates table file content. Returns list for (map display (list ..))"
        (list
            "primitive " (udp-name table) (udp-outputs table) (udp-inputs table)
            "\n  output " (udp-outputs table) ";"
            "\n  input " (udp-inputs table) ";"
            "\n  table "
            "\n  // " (udp-inputs table) " : " (udp-outputs)
            ;; line by line
            "\n  endtable"
            "\nendprimitive"
        )
    )

;;  ===================================================================
;;                  END OF R7RS LIBRARY
;;  ===================================================================
  )
)
